### DISCUSSION TITLE
拼音输入法弹框里面显示由哪部分代码控制？

### DISCUSSION THREAD

**Discussion by guyuming76** [fcitx/fcitx5#686] [Q&A] [answered]
之前我在dwl(wayland)里面用fcitx5拼音输入法，弹出选词框会显示两行，第一行显示输入的键盘字符，第二行显示候选中文
![20220908_16h12m06s_grim](https://user-images.githubusercontent.com/12045548/207605044-20bec90e-7a8d-4f13-93aa-a51963eccca6.png)

现在，在alacritty 里，弹出框只显示一行中文，而alacritty 光标处并不显示当前输入的键盘字符，但在foot里，虽然弹出框也只显示一行候选中文，但foot输入光标处会显示当前输入的字符
https://user-images.githubusercontent.com/12045548/207598411-9c2d55d3-57f5-4585-b0ed-47de70d49a3e.png

我最近升级了系统，但我不知道是那部分的升级，抑或是配置的改动造成了显示的不同。我想知道这个显示逻辑是 fcitx5 ,还是某个addon,还是虚拟终端控制的？抑或是wayland compositor控制的？

**Comment by wengxt** [comment, marked as answer]
我不知道你的 alacritty 用的是 xwayland 还是 wayland

xwayland 的话可能是受到这个 bug 的影响，建议更新。
https://github.com/rust-windowing/winit/issues/2448

wayland 的话我以为他们应该加入了 preedit 的支持来着，因为毕竟 wayland 并不支持汇报是否可以用 preedit

**Reply by guyuming76** [reply]
alacritty 用的是wayland, 测试时wayland compositor dwl 没有编入xwayland支持，进程列表里也没有xwayland进程，但是有没有可能alacritty 会通过"包裹在dwl外面"的xorg 进程访问到服务与那个xorg进程的 fcitx5进程呢？他们是如何隔离的？有的时候我在"外面的xorg"里运行emacs,结果窗口就显示到嵌套的dwl里了。

我升级了一下alacritty, 从0.10.1到gentoo包9999（github上看已经有0.11.*）了，结果表现就和foot一样了，我估计就是你说的加入了可以用preedit 支持

**Reply by wengxt** [reply]
你说的情况是不可能的。alacritty 的窗口的协议和它使用的输入法协议是一致的，如果是 x 窗口那只能是 xim，如果是 wayland 窗口那只能是 text input，你观察具体窗口是用 x 还是 wayland 就知道用的是哪个协议了

**Reply by guyuming76** [reply]
感觉fcitx5哪里变了:虚拟终端显示 preedit, 似乎好处是anthy可以显示输入键对应的假名 https://github.com/fcitx/fcitx5-anthy/issues/6 ; 然后弹出框只显示候选词，不显示preedit。wlroots, dwl 最近的升级导致fcitx5 输入拼音时报错(https://github.com/djpohly/dwl/issues/358) 有可能和这些变更有关吗？

**Reply by wengxt** [reply]
不是 fcitx5 变了，是 alacritty 变了。

至于报错，谁 crash 找谁的问题。截图里明显 crash 的是 dwl。
